stdlib/public/core/StringWordBreaking.swift

```diff
@@ -0,0 +1,690 @@
+        return (scalar, $0 &+ len)
+      }
+    }
+  }
+
+  internal func _foreignOpaqueNextWordIndex(startingAt i: Int) -> Int {
+#if _runtime(_ObjC)
+    return nextWordBoundary(startingAt: i) {
+      _internalInvariant($0 >= 0)
+
+      guard $0 < count else {
+        return nil
+      }
+
+      let scalars = String.UnicodeScalarView(self)
+      let idx = String.Index(_encodedOffset: $0)
+
+      let scalar = scalars[idx]
+      let nextIndex = scalars.index(after: idx)
+
+      return (scalar, nextIndex._encodedOffset)
+    }
+#else
+    fatalError("No foreign strings on this platform in this version of Swift.")
+#endif
+  }
+
+  internal func _opaquePreviousWordIndex(endingAt i: Int) -> Int {
+    if _slowPath(isForeign) {
+      return _foreignOpaquePreviousWordIndex(endingAt: i)
+    }
+
+    return withFastUTF8 { utf8 in
+      previousWordBoundary(endingAt: i) {
+        _internalInvariant($0 <= count)
+
+        guard $0 > 0 else {
+          return nil
+        }
+
+        let (scalar, len) = _decodeScalar(utf8, endingAt: $0)
+        return (scalar, $0 &- len)
+      }
+    }
+  }
+
+  @inline(never)
+  internal func _foreignOpaquePreviousWordIndex(endingAt i: Int) -> Int {
+    #if _runtime(_ObjC)
+    return previousWordBoundary(endingAt: i) {
+      _internalInvariant($0 <= count)
+
+      guard $0 > 0 else {
+        return nil
+      }
+
+      let scalars = String.UnicodeScalarView(self)
+      let idx = String.Index(_encodedOffset: $0)
+
+      let previousIndex = scalars.index(before: idx)
+      let scalar = scalars[previousIndex]
+
+      return (scalar, previousIndex._encodedOffset)
+    }
+#else
+    fatalError("No foreign strings on this platform in this version of Swift.")
+#endif
+  }
+}
+
+internal enum _WordQuestion {
+  case checkingRegionalIndicator(count: Int, previousRIIndex: Int)
+  case requireAHLetter
+  case requireNumeric
+  case requireHebrewLetter
+}
+
+extension _WordQuestion: Equatable {}
+
+internal struct _WordBreakingState {
+  var constraint: (question: _WordQuestion, index: Int)? = nil
+
+  var index: Int
+
+  var previousIndex: Int? = nil
+  var previousProperty: Unicode._WordBreakProperty? = nil
+
+  // When walking forward in a string, we need to not break on emoji flag
+  // sequences. Emoji flag sequences are composed of 2 regional indicators, so
+  // when we see our first (.regionalIndicator, .regionalIndicator) decision,
+  // we need to know to return false in this case. However, if the next scalar
+  // is another regional indicator, we reach the same decision rule, but in this
+  // case we actually need to break there's a boundary between emoji flag
+  // sequences.
+  var shouldBreakRI = false
+}
+
+extension _StringGuts {
+  // Returns the stride of the next word at the previous boundary offset.
+  internal func nextWordBoundary(
+    startingAt index: Int,
+    nextScalar: (Int) -> (scalar: Unicode.Scalar, end: Int)?
+  ) -> Int {
+    var (scalar, index) = nextScalar(index)!
```

Comment:
It seems like it would be clearer to precondition index is not end before invoking the closure